File: java-kata/src/main/java/com/leetcode/ValidPalindrome.java
package com.leetcode;

/**
 * <h3><a href="https://leetcode.com/problems/valid-palindrome/">Valid Palindrome</a></h3>
 * <p>
 *     A phrase is a <strong>palindrome</strong> if, after converting all uppercase letters into lowercase letters and
 *     removing all non-alphanumeric characters, it reads the same forward and backward.
 *     Alphanumeric characters include letters and numbers.
 * </p>
 * <p>
 *     Given a string <code>s</code>, return <code>true</code> if it is a palindrome, or <code>false</code> otherwise.
 * </p>
 * <p>
 *     <h6>Constraints:</h6>
 *     <ul>
 *         <li><code>1 <= s.length <= 2 * 10^5</code></li>
 *         <li><code>s</code> consists only of printable ASCII characters.</li>
 *     </ul>
 * </p>
 */
public class ValidPalindrome {
    public boolean isPalindrome(String s) {
        var left = 0;
        var right = s.length() - 1;

        while (left < right) {
            char leftChar = s.charAt(left);
            while (!Character.isLetterOrDigit(leftChar) && ++left <= right) {
                leftChar = s.charAt(left);
            }

            char rightChar = s.charAt(right);
            while (!Character.isLetterOrDigit(rightChar) && left <= --right) {
                rightChar = s.charAt(right);
            }

            if (Character.toLowerCase(leftChar) != Character.toLowerCase(rightChar)) {
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/ValidPalindrome.kt
package com.leetcode

/**
 * ### [Valid Palindrome](https://leetcode.com/problems/valid-palindrome/)
 *
 * A phrase is a **palindrome** if, after converting all uppercase letters into lowercase letters and
 * removing all non-alphanumeric characters, it reads the same forward and backward.
 * Alphanumeric characters include letters and numbers.
 *
 * Given a string `s`, return `true` if it is a palindrome, or `false` otherwise.
 *
 * ###### Constraints:
 *  - `1 <= s.length <= 2 * 10^5`
 *  - `s` consists only of printable ASCII characters.
 */
class ValidPalindrome {
    fun isPalindrome(s: String): Boolean {
        var left = 0
        var right = s.length - 1

        while (left < right) {
            var leftCh = s[left]
            while (!leftCh.isLetterOrDigit() && ++left <= right) {
                leftCh = s[left]
            }

            var rightCh = s[right]
            while (!rightCh.isLetterOrDigit() && left <= --right) {
                rightCh = s[right]
            }

            if (!leftCh.equals(rightCh, ignoreCase = true)) {
                return false
            }

            left++
            right--
        }

        return true
    }
}
